docs: 3.6 redo tutorial (#20322)
juju/juju: docs/tutorial/index.md

(tutorial)=
# Get started with Juju

Juju is an application orchestration tool for charmed applications -- that is, applications whose installation, configuration, integration, scaling, upgrading, etc., operations have been defined in a way that Juju can interpret in software packages called 'charms'.

In this tutorial you will get acquainted with Juju and charms by deploying a chat service.

> <!-- Imagine your business needs a chat service such as Mattermost backed up by a database such as PostgreSQL. In a traditional setup, this can be quite a challenge, but with Juju you'll find yourself deploying, configuring, scaling, integrating, etc., applications in no time. Let's get started! -->

Reviewer: Delete me.

Author: Done.

  System load:  0.24              Processes:             145
  Usage of /:   3.3% of 47.39GB   Users logged in:       0
  Memory usage: 3%                IPv4 address for ens3: 10.238.98.204
  Swap usage:   0%

Expanded Security Maintenance for Applications is not enabled.

18 updates can be applied immediately.
18 of these updates are standard security updates.
To see these additional updates run: apt list --upgradable

Enable ESM Apps to receive additional future security updates.
See https://ubuntu.com/esm or run: sudo pro status

To run a command as administrator (user "root"), use "sudo <command>".
See "man sudo_root" for details.

```

Anything you type after the VM shell prompt will run on the VM.

```{dropdown} Tips for usage

At any point:
- To exit the shell, press {kbd}`mod` + {kbd}`C` (e.g., {kbd}`Ctrl`+{kbd}`C`) or type `exit`.
- To stop the VM after exiting the VM shell, run `multipass stop my-juju-vm`.
- To restart the VM and re-open a shell into it, type `multipass shell my-juju-vm`.

```
```{dropdown} Tips for troubleshooting

If the VM launch fails, run `multipass delete --purge my-juju-vm` to clean up, then try the launch line again.

```

## Set up Juju

> <!-- Juju is a distributed system consisting of a controller that contains the API server and database and which coordinates with clients and agents; to have a functional Juju you need to set up at least the `juju` CLI client and a Juju controller. -->

Reviewer: Heh? Probably worth just deleting this comments.

To run a command as administrator (user "root"), use "sudo <command>".
See "man sudo_root" for details.

```

Anything you type after the VM shell prompt will run on the VM.

```{dropdown} Tips for usage

At any point:
- To exit the shell, press {kbd}`mod` + {kbd}`C` (e.g., {kbd}`Ctrl`+{kbd}`C`) or type `exit`.
- To stop the VM after exiting the VM shell, run `multipass stop my-juju-vm`.
- To restart the VM and re-open a shell into it, type `multipass shell my-juju-vm`.

```
```{dropdown} Tips for troubleshooting

If the VM launch fails, run `multipass delete --purge my-juju-vm` to clean up, then try the launch line again.

```

## Set up Juju

<!-- Juju is a distributed system consisting of a controller that contains the API server and database and which coordinates with clients and agents; to have a functional Juju you need to set up at least the `juju` CLI client and a Juju controller. -->

```{figure} tutorial-setup.svg
   :alt: Juju consists of a client and a controller and both need access to a cloud and to Charmhub

   _Juju consists of at least a client and a controller, and both need access to a cloud and to Charmhub._
```

> The way Juju works is that you use a client to talk to a controller; the controller talks to {ref}`clouds <list-of-supported-clouds>` to provision infrastructure and to [Charmhub](https://charmhub.io/) to get charms to deploy, configure, integrate, scale, upgrade, etc., applications on that infrastructure; and the controller itself must live on a cloud resource, so before you do any of that you must use the client to talk to a cloud and Charmhub to bootstrap the controller into the cloud. Let's go ahead and prepare all those pieces and make sure your Juju is good to go!

Author: Updated to "Charmhub (or a local source for charms)"

### Set up the `juju` CLI client

In Juju a (user-facing) client is anything that can talk to a Juju controller. That currently includes the `juju` CLI, the `terraform` CLI when used with the `juju` provider plugin, two Python clients, and a JavaScript client. However, to get a Juju controller you need the `juju` CLI client and it must have access to a cloud and Charmhub. Let's go ahead and set it up!

In your VM, install the `juju` CLI client:

```{terminal}
:user: ubuntu
:host: my-juju-vm
:input: sudo snap install juju

juju (3/stable) 3.6.8 from Canonical✓ installed
```

Now, ensure the client has access to your cloud (i.e., knows where to find your cloud and has the credentials to access your cloud). For a localhost MicroK8s cloud installed from a strictly confined snap like ours, your `juju` client can read the local kubeconfig file and retrieve the cloud definition (and credentials) from there automatically, as you can verify:

```{terminal}
:user: ubuntu
:host: my-juju-vm
:input: juju clouds --client

# Output should show microk8s.
# If it doesn't: Exit the VM (`exit`), re-enter it (`multipass shell my-juju-vm`), then try again.

```

> Ensure also that the client has access to Charmhub by launching a random query, e.g., "ingress", and asking for more information about one of the results it shows:

Author: This is about the keyword used for `juju find`. I'm trying to prove the client can access Charmhub by showing something such as `juju find` works.

```{terminal}
:user: ubuntu
:host: my-juju-vm
:input: juju find ingress

# Output should show all the charms or charm bundles related to this query that are available on Charmhub.
# For best results always double-check Charmhub.

:input: juju info traefik-k8s

# Output should show name, publisher, quick description, integration endpoints, etc.
# For the full features and more info see directly the charm's page on Charmhub.
```

Our `juju` client is ready! Sneak a peek at what it can do (`juju help commands` -- you can pipe a query to zoom in on feature: `juju help commands | grep application`). Check status (`juju status` and `juju debug-log` -- not much to see there yet). Keep going...

```{tip}

Split your terminal window into 3 (or open 3 terminal windows). In all, access your Multipass VM shell (`multipass shell my-juju-vm`) and then:

**Shell 1:** Keep using it as you've already been doing so far, namely to type the commands in this tutorial.

> **Shell 2:** Run `juju status --relations --color --watch 1s` to watch your deployment status evolve. Things are all right if your `App Status` and your `Unit - Workload` reach `active` and your `Unit - Agent` reaches `idle`. To exit and return to the terminal prompt, press {kbd}`mod` + {kbd}`C` (e.g., {kbd}`Ctrl`+{kbd}`C`).

Reviewer 2: ```suggestion
**Shell 2:** Run `watch --color juju status --color` to watch your deployment status evolve. Things are all right if your `App Status` and your `Unit - Workload` reach `active` and your `Unit - Agent` reaches `idle`. To exit and return to the terminal prompt, press {kbd}`mod` + {kbd}`C` (e.g., {kbd}`Ctrl`+{kbd}`C`).
```

This is very much 3.x ism, we dropped the `--watch` flag in 4.0. So with one change, you could make it both 3.x and 4.x.

`watch --color juju status --color `

Also, it might be worth raising a bug against juju about the fact that we need the second `--color`. We should be able to detect that it can use color.

> **Shell 3:** Run `juju debug-log` to watch all the details behind your deployment status. (Especially useful when things don't evolve as expected. In that case, please get in touch.)

Reviewer 2: ```suggestion
**Shell 3:** Run `juju debug-log -m controller` to watch all the details behind your deployment status. (Especially useful when things don't evolve as expected. In that case, please get in touch.)
```

The logs for the model doesn't really tell you anything, you want the controller logs. This changes in 4.x, but for 3.x controller is very much king.

```

### Set up a Juju controller

> A Juju controller is your Juju control plane -- the entity that holds the Juju API server and Juju's internal database. Anything you do in Juju post-controller-setup goes through a Juju controller, and to work properly the controller needs access to a cloud and to Charmhub (or a local source of charms). Let's set it up!

Reviewer 2: ```suggestion
A Juju controller is your Juju control plane -- the entity that holds the Juju API server and Juju's database. Anything you do in Juju post-controller-setup goes through a Juju controller, and to work properly the controller needs access to a cloud and to Charmhub (or a local source of charms). Let's set it up!
```

  1. a 'controller' namespace (`kubectl get namespaces`) holding
  1. a 'controller-0' pod (`kubectl -n controller-my-first-juju-controller get pods`) holding
  1. 3 containers: 'api-server', 'charm', 'mongodb' (`kubectl -n controller describe pod controller-0`).
- On the Juju side (in the Juju controller's database) you've now got:
  1. a Juju 'my-first-juju-controller' controller (i.e., control plane), with access to the 'microk8s' cloud (`juju show-controller my-first-juju-controller`), and holding
  1. a Juju model (i.e., workspace; corresponding to a Kubernetes namespace) implicitly called 'controller', using resources from the 'microk8s' cloud (`juju switch controller`, then `juju show-model controller`), and  holding
  1. a Juju application (i.e., deployed charm -- in this case, the `juju-controller` charm) called 'controller' with a Juju unit (i.e., replica; corresponding to a Kubernetes pod) called 'controller/0', deployed on a resource from the 'microk8s' cloud (`juju status`), and holding
  1. the same 3 containers we saw on the MicroK8s side (`juju ssh --container [api-server | charm | mongodb] 0`), which hold the various bits of a running Juju controller:

>     - 'charm': contains a Juju unit agent and `juju-controller` charm code. The Juju unit agent pings the Juju controler agent (see next item) to check the unit's target state against the state in the database (see two items down) and progresses the unit's state by executing the charm code.

Reviewer 2: ```suggestion
    - 'charm': contains a Juju unit agent and `juju-controller` charm code. The Juju unit agent sends requests to the Juju controller agent (see next item) to check the unit's target state against the state in the database (see two items down) and progresses the unit's state by executing the charm code.
```

Ping means something different, and I don't want to mix terminology here.

Author: Updated, thanks!

    - 'api-server': contains Pebble, a lightweight process supervisor, and the Juju controller agent, which also wraps the Juju API server. The controller agent reads and writes to the database (see next item).
    - 'mongodb': contains Juju's internal database.

These details are specific to Kubernetes, bootstrap, and the controller application, but this is fundamentally how Juju operates for all clouds, all operations, and all applications: the cloud provides a resource and Charmhub provides charms, and the result is a resource that holds the Juju unit agent and the charm code, where the Juju unit agent always pings the Juju controller agent to reconcile state, then executes the charm code accordingly, and the charm reacts as it's been programmed to react, e.g., by installing a workload.

> Back to our main concern, to be fully operational a controller needs access to a cloud and to Charmhub. We saw just now that our controller already has access to our 'microk8s' cloud -- this access was granted implicitly through bootstrap. Also, as before, so long as you're connected to the internet, your controller has access to Charmhub too. Your controller is all set!

Reviewer 2: > to be fully operational a controller needs access to a cloud and to Charmhub.

That's only true for this tutorial. You can have air gapped setups that doesn't require Charmhub at all i.e. you can make charms without ever using Charmhub i.e. you can bootstrap with a local charm.

```
juju bootstrap lxd test --controller-charm-path=./juju-controller_r117.charm
```

Now, its dependencies. Mattermost needs a PostgreSQL database, and [its charmed version supports an easy way to integrate with such a database](https://charmhub.io/mattermost-k8s/integrations#db). Let's deploy [PostgreSQL](https://charmhub.io/postgresql-k8s) in the recommended way, from track 14 with risk `stable`; with `--trust` -- i.e., permission to use our cloud credentials (this charm needs to create and manage some Kubernetes resources); because we're just playing around, setting [the `profile` config](https://charmhub.io/postgresql-k8s/configurations#profile) to `testing`, so we don't use too many resources; and, just for fun, with `-n 2`, that is, two replicas (in a real life setting you'll want to distribute them over multiple nodes -- something Juju would do automatically here too, except we're doing everything on a single node).

```{terminal}
:user: ubuntu
:host: my-juju-vm
:input: juju deploy postgresql-k8s --channel 14/stable --trust --config profile=testing -n 2
Deployed "postgresql-k8s" from charm-hub charm "postgresql-k8s", revision 495 in channel 14/stable on ubuntu@22.04/stable
```

> Mattermost wants PostgreSQL status to be TLS-encrypted. There are a few ways to do that. Because we're just trying things out, we can use [Self Signed X.509 Certificates](https://charmhub.io/self-signed-certificates) (don't do this in production!). Let's deploy it and integrate it with our PostgreSQL to enable TLS encryption on our PostgreSQL cluster:

Reviewer 2: nit: I wonder how hard it is to use let's encrypt. I wonder if it'll be roughly the same as self-signed and without the scary `(don't do this in production!)`.

Author: If you don't mind, I'll defer this to future work. 

For context: We already know our current setup isn't great -- e.g., Massi tells me we should switch to Canonical K8s and I hear the Mattermost charm isn't very well maintained, so those are all content changes we'll want to make some time soon. In this PR I was mostly just focusing on the too much and too little, and the "don't use it in production" disclaimer was there mainly because someone asked for it. Truth be told, nobody should be using our tutorial setup exactly as is in production. E.g., what I suggest for scaling Mattermost and PostgreSQL is mostly just an excuse to showcase vertical and horizontal scaling, and it might not necessarily make sense from an operations perspective. (PS I did try to ask around what would make sense but people hedged a lot as it's ultimately at the discretion of the SRE.)
